File: connector/src/main/scala/com/google/cloud/spark/bigquery/ArrowBinaryIterator.scala
/*
 * Copyright 2018 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.cloud.spark.bigquery

import java.io.{ByteArrayInputStream, SequenceInputStream}

import com.google.protobuf.ByteString
import org.apache.arrow.memory.RootAllocator
import org.apache.arrow.vector.VectorSchemaRoot
import org.apache.arrow.vector.ipc.{ArrowReader, ArrowStreamReader}
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.vectorized.{ColumnVector, ColumnarBatch}

import collection.JavaConverters._

/**
 * An iterator for scanning over rows serialized in Arrow format
 * @param columnsInOrder Sequence of columns in the schema
 * @param schema Schema in arrow format
 * @param rowsInBytes Rows serialized in binary format for Arrow
 */
class ArrowBinaryIterator(columnsInOrder: Seq[String],
                          schema: ByteString,
                          rowsInBytes: ByteString) extends Iterator[InternalRow] {

  val allocator = ArrowBinaryIterator.rootAllocator.newChildAllocator("ArrowBinaryIterator",
    0, ArrowBinaryIterator.maxAllocation)

  val bytesWithSchemaStream = new SequenceInputStream(
    new ByteArrayInputStream(schema.toByteArray),
    new ByteArrayInputStream(rowsInBytes.toByteArray))

  val arrowStreamReader = new ArrowStreamReader(bytesWithSchemaStream, allocator)

  val arrowReaderIterator = new ArrowReaderIterator(arrowStreamReader)
  val iterator = arrowReaderIterator.flatMap(root => toArrowRows(root, columnsInOrder))

  private def toArrowRows(root: VectorSchemaRoot, namesInOrder: Seq[String]):
  Iterator[InternalRow] = {
    val vectors = namesInOrder.map(root.getVector)

    val columns = vectors.map { vector =>
      try {
        new ArrowSchemaConverter(vector).asInstanceOf[ColumnVector]
      } catch {
        case e : UnsupportedOperationException =>
          throw new Exception(vector.getClass.getName, e)
      }
    }.toArray

    val batch = new ColumnarBatch(columns)
    batch.setNumRows(root.getRowCount)
    batch.rowIterator().asScala
  }

  class ArrowReaderIterator(reader: ArrowReader) extends Iterator[VectorSchemaRoot] {
    var closed = false
    var current: VectorSchemaRoot = null

    def ensureClosed(): Unit = {
      if (!closed) {
        reader.close()
        closed = true
      }
    }

    override def hasNext: Boolean = {
      if (current != null) {
        return true
      }
      val res = reader.loadNextBatch()
      if (res) {
        current = reader.getVectorSchemaRoot
      } else {
        ensureClosed()
      }
      res
    }

    override def next(): VectorSchemaRoot = {
      assert(current != null)
      val res = current
      current = null
      res
    }
  }

  override def hasNext: Boolean = iterator.hasNext

  override def next(): InternalRow = iterator.next()
}

object ArrowBinaryIterator {
  // max allocation value for the allocator
  var maxAllocation = Long.MaxValue
  val rootAllocator = new RootAllocator(maxAllocation)
}

File: connector/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java
package com.google.cloud.spark.bigquery;

import com.google.common.collect.ImmutableList;
import com.google.protobuf.ByteString;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.apache.arrow.vector.ipc.ArrowStreamReader;
import org.apache.spark.sql.catalyst.InternalRow;
import org.apache.spark.sql.vectorized.ColumnVector;
import org.apache.spark.sql.vectorized.ColumnarBatch;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.SequenceInputStream;
import java.io.UncheckedIOException;
import java.util.*;
import java.util.stream.Collectors;

public class ArrowBinaryIterator implements Iterator<InternalRow> {

    private static long maxAllocation = Long.MAX_VALUE;
    ArrowReaderIterator arrowReaderIterator;
    Iterator<InternalRow> currentIterator;
    List<String> columnsInOrder;

    public ArrowBinaryIterator(List<String> columnsInOrder, ByteString schema, ByteString rowsInBytes) {
        BufferAllocator allocator = (new RootAllocator(maxAllocation)).newChildAllocator("ArrowBinaryIterator",
                0, maxAllocation);

        SequenceInputStream bytesWithSchemaStream = new SequenceInputStream(
                new ByteArrayInputStream(schema.toByteArray()),
                new ByteArrayInputStream(rowsInBytes.toByteArray()));

        ArrowStreamReader arrowStreamReader = new ArrowStreamReader(bytesWithSchemaStream, allocator);
        arrowReaderIterator = new ArrowReaderIterator(arrowStreamReader);
        currentIterator = ImmutableList.<InternalRow>of().iterator();
        this.columnsInOrder = columnsInOrder;
    }

    @Override
    public boolean hasNext() {
        while (!currentIterator.hasNext()) {
            if (!arrowReaderIterator.hasNext()) {
                return false;
            }
            currentIterator = toArrowRows(arrowReaderIterator.next(), columnsInOrder);
            arrowReaderIterator.next();
        }

        return currentIterator.hasNext();
    }

    @Override
    public InternalRow next() {
        return currentIterator.next();
    }

    private Iterator<InternalRow> toArrowRows(VectorSchemaRoot root, List<String> namesInOrder) {
        List<FieldVector> vectors = namesInOrder.stream().map(name -> root.getVector(name)).collect(Collectors.toList());
        ColumnVector[] columns = vectors.stream().map(vector -> new ArrowSchemaConverter(vector))
                .collect(Collectors.toList()).toArray(new ColumnVector[0]);

        ColumnarBatch batch = new ColumnarBatch(columns);
        batch.setNumRows(root.getRowCount());
        return batch.rowIterator();
    }
}

class ArrowReaderIterator implements Iterator<VectorSchemaRoot> {

    boolean closed = false;
    VectorSchemaRoot current = null;
    ArrowReader reader;
    private static final Logger log = LoggerFactory.getLogger(AvroBinaryIterator.class);

    public ArrowReaderIterator(ArrowReader reader) {
        this.reader = reader;
    }

    @Override
    public boolean hasNext() {
        if (current != null) {
            return true;
        }

        try {
            boolean res = reader.loadNextBatch();
            if (res) {
                current = reader.getVectorSchemaRoot();
            } else {
                ensureClosed();
            }
            return res;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public VectorSchemaRoot next() {
        VectorSchemaRoot res = current;
        current = null;
        return res;
    }

    private void ensureClosed() throws IOException {
        if (!closed) {
            reader.close();
            closed = true;
        }
    }
}
